engine-tests/src/test/java/org/terasology/engine/rendering/world/ChunkMeshWorkerTest.java

```diff
@@ -23,107 +29,158 @@
 public class ChunkMeshWorkerTest {
     static Vector3ic position0 = new Vector3i(123, 456, 789);
 
+    final Vector3i currentPosition = new Vector3i(position0);
+
     ChunkMeshWorker worker;
+    Comparator<RenderableChunk> comparator;
+    StepVerifier.FirstStep<Chunk> verifier;
 
-    static Optional<Tuple2<Chunk, ChunkMesh>> alwaysCreateMesh(Chunk chunk) {
-        return Optional.of(Tuples.of(chunk, mock(ChunkMesh.class)));
+    static Mono<Tuple2<Chunk, ChunkMesh>> alwaysCreateMesh(Chunk chunk) {
+        chunk.setDirty(false);
+        return Mono.just(Tuples.of(chunk, mock(ChunkMesh.class)));
     }
 
     @BeforeEach
     void makeWorker() {
-        worker = new ChunkMeshWorker(ChunkMeshWorkerTest::alwaysCreateMesh, null);
+        comparator = Comparator.comparingDouble(chunk ->
+                chunk.getRenderPosition().distanceSquared(currentPosition.x, currentPosition.y, currentPosition.z)
+        );
+
+        StepVerifier.setDefaultTimeout(Duration.ofSeconds(10));
+
+        // Use virtual time so we don't have to wait around in real time
+        // to see whether there are more events pending.
+        // Requires that the schedulers be created _inside_ the withVirtualTime supplier.
+        verifier = StepVerifier.withVirtualTime(() -> {
+            worker = new ChunkMeshWorker(
+                    ChunkMeshWorkerTest::alwaysCreateMesh,
+                    comparator,
+                    Schedulers.parallel(),
+                    Schedulers.single()
+            );
+            return worker.getCompletedChunks();
+        });
     }
 
-    @Test
-    void testChunkIsNotProcessedTwice() {
-        var chunk1 = new DummyChunk(position0);
-        var chunk2 = new DummyChunk(position0);
-        worker.add(chunk1);
-        worker.add(chunk2);
-        // drain
-        fail("TODO: assert number of results == 1");
+    static Chunk newDirtyChunk(Vector3ic position) {
+        var chunk = new DummyChunk(position);
+        chunk.markReady();
+        chunk.setDirty(true);
+        return chunk;
     }
 
     @Test
-    void testChunkIsRegeneratedIfDirty() {
-        var chunk = new DummyChunk(position0);
-        worker.add(chunk);
-        // tick - work function has started
-        chunk.setDirty(true);
-        worker.add(chunk);
-        // drain
-        fail("TODO: assert number of results == 2");
+    void testMultipleChunks() {
+        var chunk1 = newDirtyChunk(position0);
+        var chunk2 = newDirtyChunk(new Vector3i(position0).add(1, 0, 0));
+
+        verifier.then(() -> {
+                    worker.add(chunk1);
+                    worker.add(chunk2);
+                    worker.update();
+                })
+                .expectNext(chunk1, chunk2)
```

Comment:
does this care about the order? if so, we could include that in the test name, e.g. `testMultipleChunksOrdered` or something similar

Reply:
This is the thing I'm most concerned with from this review.

I came up with something that's probably technically correct, but the readability is horrible. Hopefully someone will answer this post with better ideas: [How to make order-independent assertions on Flux output?](https://stackoverflow.com/q/72106373/9585)

Reply:
The author answered my post and I've made some changes accordingly.

I'm torn, because I think using the `getChunksThatResultFrom` helper I added works well for these cases where you just want the final output and examine all if it together.

And I think StepVerifier is the right sort of tool to use for some of these cases where we're worried about how things change if you add more work while other things are still in progress.

But having some tests written one way and some tests written the other, and having these two different ways of getting a ChunkMeshWorker—I don't feel good about it.

I think, though, that after I add some documentation for the new function, it'll be Close Enough.

The goal here was to get tests that can keep us in line as we rework this code, and I expect the details of how we test will change as we do that.